drivers/display/display_sdl.c

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright (c) 2018 Jan Van Winkel <jan.van_winkel@dxplore.eu>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <display.h>
+
+#include <SDL.h>
+#include <string.h>
+
+#define LOG_LEVEL CONFIG_LOG_SDL_DISPLAY_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(display);
+
+struct sdl_display_data {
+	SDL_Window *window;
+	SDL_Renderer *renderer;
+	SDL_Texture *texture;
+	bool display_on;
+	enum display_pixel_format current_pixel_format;
+};
+
+static struct sdl_display_data sdl_display_data;
+
+static int sdl_display_init(struct device *dev)
+{
+	struct sdl_display_data *dis_data =
+	    (struct sdl_display_data *)dev->driver_data;
+	LOG_DBG("Initializing display driver");
+
+	memset(dis_data, 0, sizeof(struct sdl_display_data));
+
+	dis_data->current_pixel_format = PIXEL_FORMAT_ARGB_8888;
+
+	dis_data->window =
+	    SDL_CreateWindow("Zephyr Display", SDL_WINDOWPOS_UNDEFINED,
+			     SDL_WINDOWPOS_UNDEFINED, CONFIG_SDL_DISPLAY_X_RES,
+			     CONFIG_SDL_DISPLAY_Y_RES, SDL_WINDOW_SHOWN);
+	if (dis_data->window == NULL) {
+		LOG_ERR("Failed to create SDL window: %s", SDL_GetError());
+		return -EIO;
+	}
+
+	dis_data->renderer =
+	    SDL_CreateRenderer(dis_data->window, -1, SDL_RENDERER_ACCELERATED);
+	if (dis_data->renderer == NULL) {
+		LOG_ERR("Failed to create SDL renderer: %s",
+			    SDL_GetError());
+		return -EIO;
+	}
+
+	dis_data->texture = SDL_CreateTexture(
+	    dis_data->renderer, SDL_PIXELFORMAT_ARGB8888,
+	    SDL_TEXTUREACCESS_STREAMING, CONFIG_SDL_DISPLAY_X_RES,
+	    CONFIG_SDL_DISPLAY_Y_RES);
+	if (dis_data->texture == NULL) {
+		LOG_ERR("Failed to create SDL texture: %s", SDL_GetError());
+		return -EIO;
+	}
+
+	dis_data->display_on = false;
+	SDL_SetRenderDrawColor(dis_data->renderer, 0, 0, 0, 0xFF);
+	SDL_RenderClear(dis_data->renderer);
+	SDL_RenderPresent(dis_data->renderer);
+
+	return 0;
+}
+
+static void sdl_display_write_argb8888(void *disp_buf, int dis_pitch,
+		const struct display_buffer_descriptor *desc, const void *buf)
+{
+	__ASSERT((4 * desc->width) == dis_pitch,
+			"Display buffer pitch not equal to buffer width");
+	__ASSERT((4 * desc->pitch * desc->height) <= desc->bu_size,
+			"Input buffer to small");
+
+	memcpy(disp_buf, buf, 4 * desc->pitch * desc->height);
+}
+
+static void sdl_display_write_rgb888(u8_t *disp_buf, int dis_pitch,
+		const struct display_buffer_descriptor *desc, const void *buf)
+{
+	u32_t w_idx;
+	u32_t h_idx;
+	u32_t pixel;
+	const u8_t *byte_ptr;
+
+	__ASSERT((4 * desc->width) == dis_pitch,
+			"Display buffer pitch not equal to buffer width");
+	__ASSERT((3 * desc->pitch * desc->height) <= desc->bu_size,
+			"Input buffer to small");
+
+	for (h_idx = 0; h_idx < desc->height; ++h_idx) {
+		for (w_idx = 0; w_idx < desc->width; ++w_idx) {
+			byte_ptr = (const u8_t *)buf +
+				3 * ((h_idx * desc->pitch) + w_idx);
+			pixel = *byte_ptr << 16;
+			pixel |= *(byte_ptr + 1) << 8;
+			pixel |= *(byte_ptr + 2);
+			*((u32_t *)disp_buf) = pixel;
+			*disp_buf += 4;
```

Comment:
do you want to add 4 to the data? or move the pointer 4 bytes ahead?

Reply:
nice catch